File: ios/Runner/PeerServicePlugin.swift
import Flutter
import Foundation
import MultipeerConnectivity

/// Bridges spec 22 `PeerService` to Apple's MultipeerConnectivity framework.
///
/// Channels (must match `lib/services/share/channel_peer_service.dart`):
/// - method `noti.peer/control` — start/stop/invite/accept/reject/sendBytes/sendFile/cancel/disconnect
/// - event  `noti.peer/peers` — list of discovered + connected peers
/// - event  `noti.peer/invites` — inbound invitations awaiting accept/reject
/// - event  `noti.peer/payloads` — received bytes / streamed files
/// - event  `noti.peer/transfers` — per-transfer progress for both directions
public final class PeerServicePlugin: NSObject {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let plugin = PeerServicePlugin()
        let control = FlutterMethodChannel(
            name: "noti.peer/control",
            binaryMessenger: registrar.messenger()
        )
        control.setMethodCallHandler(plugin.handle)
        plugin.peersChannel = FlutterEventChannel(
            name: "noti.peer/peers", binaryMessenger: registrar.messenger())
        plugin.invitesChannel = FlutterEventChannel(
            name: "noti.peer/invites", binaryMessenger: registrar.messenger())
        plugin.payloadsChannel = FlutterEventChannel(
            name: "noti.peer/payloads", binaryMessenger: registrar.messenger())
        plugin.transfersChannel = FlutterEventChannel(
            name: "noti.peer/transfers", binaryMessenger: registrar.messenger())
        plugin.peersChannel?.setStreamHandler(plugin.peersHandler)
        plugin.invitesChannel?.setStreamHandler(plugin.invitesHandler)
        plugin.payloadsChannel?.setStreamHandler(plugin.payloadsHandler)
        plugin.transfersChannel?.setStreamHandler(plugin.transfersHandler)
    }

    // Channels
    private var peersChannel: FlutterEventChannel?
    private var invitesChannel: FlutterEventChannel?
    private var payloadsChannel: FlutterEventChannel?
    private var transfersChannel: FlutterEventChannel?
    fileprivate let peersHandler = SinkHandler()
    fileprivate let invitesHandler = SinkHandler()
    fileprivate let payloadsHandler = SinkHandler()
    fileprivate let transfersHandler = SinkHandler()

    // MultipeerConnectivity state — recreated on each start().
    private var session: MCSession?
    private var advertiser: MCNearbyServiceAdvertiser?
    private var browser: MCNearbyServiceBrowser?
    private var localPeerID: MCPeerID?
    private var serviceType: String = "noti-share"

    private var peersById: [String: MCPeerID] = [:]
    private var stateById: [String: String] = [:]

    private var pendingInvites: [String: (Bool, MCSession?) -> Void] = [:]
    private var inviteToPeer: [String: MCPeerID] = [:]

    private var sendProgress: [String: Progress] = [:]
    private var receiveProgress: [String: Progress] = [:]
    private var sendObservers: [String: NSKeyValueObservation] = [:]
    private var receiveObservers: [String: NSKeyValueObservation] = [:]
    private var sendTransferIdByPeer: [String: String] = [:]
    /// Maps `"<peerId>|<resourceName>"` to the transferId we minted on the
    /// receive side. Replaces the previous `objc_setAssociatedObject` trick,
    /// which used a single file-scope key and could mis-attribute simultaneous
    /// transfers.
    private var receiveTransferIds: [String: String] = [:]

    // MARK: - Method dispatch

    fileprivate func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start": start(call.arguments, result: result)
        case "stop": stop(result: result)
        case "invite": invite(call.arguments, result: result)
        case "acceptInvite": acceptInvite(call.arguments, result: result)
        case "rejectInvite": rejectInvite(call.arguments, result: result)
        case "sendBytes": sendBytes(call.arguments, result: result)
        case "sendFile": sendFile(call.arguments, result: result)
        case "cancelTransfer": cancelTransfer(call.arguments, result: result)
        case "disconnect": disconnect(call.arguments, result: result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    private func start(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any],
              let role = map["role"] as? String,
              let displayName = map["displayName"] as? String,
              let serviceTypeArg = map["serviceType"] as? String
        else {
            result(FlutterError(code: "bad_args", message: "start: missing arguments", details: nil))
            return
        }
        if session != nil {
            result(FlutterError(code: "already_active", message: "PeerService already running", details: nil))
            return
        }

        serviceType = serviceTypeArg
        let peerID = MCPeerID(displayName: displayName.isEmpty ? "noti" : displayName)
        let session = MCSession(peer: peerID, securityIdentity: nil, encryptionPreference: .required)
        session.delegate = self
        self.session = session
        self.localPeerID = peerID

        if role == "advertise" || role == "both" {
            let adv = MCNearbyServiceAdvertiser(peer: peerID, discoveryInfo: nil, serviceType: serviceTypeArg)
            adv.delegate = self
            adv.startAdvertisingPeer()
            advertiser = adv
        }
        if role == "discover" || role == "both" {
            let br = MCNearbyServiceBrowser(peer: peerID, serviceType: serviceTypeArg)
            br.delegate = self
            br.startBrowsingForPeers()
            browser = br
        }
        emitPeers()
        result(nil)
    }

    private func stop(result: @escaping FlutterResult) {
        // Invalidate KVO observers BEFORE clearing the progress dictionaries,
        // otherwise a callback already queued on the main thread can fire
        // against a half-cleared state.
        sendObservers.values.forEach { $0.invalidate() }
        receiveObservers.values.forEach { $0.invalidate() }
        sendObservers.removeAll()
        receiveObservers.removeAll()

        advertiser?.delegate = nil
        browser?.delegate = nil
        advertiser?.stopAdvertisingPeer()
        browser?.stopBrowsingForPeers()
        session?.delegate = nil
        session?.disconnect()

        for (_, p) in sendProgress { p.cancel() }
        for (_, p) in receiveProgress { p.cancel() }
        sendProgress.removeAll()
        receiveProgress.removeAll()
        sendTransferIdByPeer.removeAll()
        receiveTransferIds.removeAll()
        peersById.removeAll()
        stateById.removeAll()
        pendingInvites.removeAll()
        inviteToPeer.removeAll()
        advertiser = nil
        browser = nil
        session = nil
        localPeerID = nil
        peersHandler.send([])
        result(nil)
    }

    private func invite(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any],
              let peerId = map["peerId"] as? String,
              let peer = peersById[peerId],
              let session = session,
              let browser = browser
        else {
            result(FlutterError(code: "no_peer", message: "Unknown peer or not browsing", details: nil))
            return
        }
        browser.invitePeer(peer, to: session, withContext: nil, timeout: 30)
        stateById[peerId] = "inviting"
        emitPeers()
        result(nil)
    }

    private func acceptInvite(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any],
              let inviteId = map["inviteId"] as? String,
              let handler = pendingInvites.removeValue(forKey: inviteId)
        else {
            result(FlutterError(code: "no_invite", message: "Unknown invite id", details: nil))
            return
        }
        handler(true, session)
        if let peer = inviteToPeer.removeValue(forKey: inviteId), let pid = peersById.first(where: { $0.value == peer })?.key {
            stateById[pid] = "accepting"
            emitPeers()
        }
        result(nil)
    }

    private func rejectInvite(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any],
              let inviteId = map["inviteId"] as? String,
              let handler = pendingInvites.removeValue(forKey: inviteId)
        else {
            result(FlutterError(code: "no_invite", message: "Unknown invite id", details: nil))
            return
        }
        handler(false, nil)
        inviteToPeer.removeValue(forKey: inviteId)
        result(nil)
    }

    private func sendBytes(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any],
              let peerId = map["peerId"] as? String,
              let peer = peersById[peerId],
              let session = session,
              let bytes = map["bytes"] as? FlutterStandardTypedData
        else {
            result(FlutterError(code: "bad_args", message: "sendBytes args", details: nil))
            return
        }
        let transferId = UUID().uuidString
        let total = bytes.data.count
        do {
            try session.send(bytes.data, toPeers: [peer], with: .reliable)
            transfersHandler.send(transferEvent(transferId: transferId, peerId: peerId, direction: "send", phase: "completed", bytes: total, total: total))
            result(transferId)
        } catch {
            transfersHandler.send(transferEvent(transferId: transferId, peerId: peerId, direction: "send", phase: "failed", bytes: 0, total: total, error: error.localizedDescription))
            result(FlutterError(code: "send_failed", message: error.localizedDescription, details: nil))
        }
    }

    private func sendFile(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any],
              let peerId = map["peerId"] as? String,
              let peer = peersById[peerId],
              let session = session,
              let path = map["path"] as? String
        else {
            result(FlutterError(code: "bad_args", message: "sendFile args", details: nil))
            return
        }
        let url = URL(fileURLWithPath: path)
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: path),
              let fileSize = attrs[.size] as? Int
        else {
            result(FlutterError(code: "no_file", message: "Cannot stat file", details: nil))
            return
        }
        if let cap = map["maxBytes"] as? Int, fileSize > cap {
            result(FlutterError(code: "too_large", message: "File exceeds size cap", details: nil))
            return
        }
        let transferId = UUID().uuidString
        let resourceName = url.lastPathComponent
        sendTransferIdByPeer["\(peerId)|\(resourceName)"] = transferId
        let progress = session.sendResource(at: url, withName: resourceName, toPeer: peer) { [weak self] err in
            guard let self = self else { return }
            let phase = err == nil ? "completed" : "failed"
            self.transfersHandler.send(self.transferEvent(
                transferId: transferId, peerId: peerId, direction: "send",
                phase: phase, bytes: fileSize, total: fileSize,
                error: err?.localizedDescription))
            self.sendObservers[transferId]?.invalidate()
            self.sendObservers.removeValue(forKey: transferId)
            self.sendProgress.removeValue(forKey: transferId)
        }
        if let progress = progress {
            sendProgress[transferId] = progress
            sendObservers[transferId] = progress.observe(\.fractionCompleted) { [weak self] p, _ in
                guard let self = self else { return }
                let bytes = Int(Double(fileSize) * p.fractionCompleted)
                self.transfersHandler.send(self.transferEvent(
                    transferId: transferId, peerId: peerId, direction: "send",
                    phase: "inProgress", bytes: bytes, total: fileSize))
            }
        }
        result(transferId)
    }

    private func cancelTransfer(_ args: Any?, result: @escaping FlutterResult) {
        guard let map = args as? [String: Any], let transferId = map["transferId"] as? String else {
            result(FlutterError(code: "bad_args", message: "cancelTransfer", details: nil))
            return
        }
        sendProgress[transferId]?.cancel()
        receiveProgress[transferId]?.cancel()
        result(nil)
    }

    private func disconnect(_ args: Any?, result: @escaping FlutterResult) {
        // Multipeer does not expose per-peer disconnect from one side without tearing the
        // whole session. This call is a stub that disconnects the entire session.
        session?.disconnect()
        result(nil)
    }

    // MARK: - Helpers

    private func peerKey(for peer: MCPeerID) -> String {
        // Deterministic id reused across browse/state cycles.
        return "\(peer.displayName)#\(peer.hash)"
    }

    fileprivate func emitPeers() {
        let list = peersById.map { (id, peer) -> [String: Any] in
            return [
                "id": id,
                "displayName": peer.displayName,
                "state": stateById[id] ?? "found",
            ]
        }
        peersHandler.send(list)
    }

    fileprivate func transferEvent(
        transferId: String, peerId: String, direction: String, phase: String,
        bytes: Int, total: Int, error: String? = nil
    ) -> [String: Any] {
        var ev: [String: Any] = [
            "transferId": transferId,
            "peerId": peerId,
            "direction": direction,
            "phase": phase,
            "bytes": bytes,
            "total": total,
        ]
        if let e = error { ev["error"] = e }
        return ev
    }
}

// MARK: - MCSessionDelegate

extension PeerServicePlugin: MCSessionDelegate {
    public func session(_ session: MCSession, peer peerID: MCPeerID, didChange state: MCSessionState) {
        let id = peerKey(for: peerID)
        peersById[id] = peerID
        switch state {
        case .notConnected: stateById[id] = "disconnected"
        case .connecting: stateById[id] = "accepting"
        case .connected: stateById[id] = "connected"
        @unknown default: stateById[id] = "found"
        }
        emitPeers()
    }

    public func session(_ session: MCSession, didReceive data: Data, fromPeer peerID: MCPeerID) {
        let id = peerKey(for: peerID)
        peersById[id] = peerID
        payloadsHandler.send([
            "peerId": id,
            "bytes": FlutterStandardTypedData(bytes: data),
        ])
    }

    public func session(
        _ session: MCSession,
        didStartReceivingResourceWithName resourceName: String,
        fromPeer peerID: MCPeerID,
        with progress: Progress
    ) {
        let id = peerKey(for: peerID)
        let transferId = UUID().uuidString
        receiveProgress[transferId] = progress
        receiveTransferIds["\(id)|\(resourceName)"] = transferId
        receiveObservers[transferId] = progress.observe(\.fractionCompleted) { [weak self] p, _ in
            guard let self = self,
                  self.receiveProgress[transferId] != nil
            else { return }
            self.transfersHandler.send(self.transferEvent(
                transferId: transferId, peerId: id, direction: "receive",
                phase: "inProgress",
                bytes: Int(Double(p.totalUnitCount) * p.fractionCompleted),
                total: Int(p.totalUnitCount)))
        }
    }

    public func session(
        _ session: MCSession,
        didFinishReceivingResourceWithName resourceName: String,
        fromPeer peerID: MCPeerID,
        at localURL: URL?,
        withError error: Error?
    ) {
        let id = peerKey(for: peerID)
        let key = "\(id)|\(resourceName)"
        let transferId = receiveTransferIds.removeValue(forKey: key) ?? UUID().uuidString
        let total = Int(receiveProgress[transferId]?.totalUnitCount ?? 0)
        let phase: String
        if let err = error {
            phase = err.localizedDescription.contains("cancel") ? "cancelled" : "failed"
        } else {
            phase = "completed"
        }
        transfersHandler.send(transferEvent(
            transferId: transferId, peerId: id, direction: "receive",
            phase: phase, bytes: total, total: total,
            error: error?.localizedDescription))
        receiveObservers[transferId]?.invalidate()
        receiveObservers.removeValue(forKey: transferId)
        receiveProgress.removeValue(forKey: transferId)
        if phase == "completed", let localURL = localURL {
            payloadsHandler.send([
                "peerId": id,
                "bytes": FlutterStandardTypedData(bytes: Data()),
                "filePath": localURL.path,
            ])
        }
    }

    public func session(_ session: MCSession, didReceive stream: InputStream, withName streamName: String, fromPeer peerID: MCPeerID) {
        // Streams are unused — spec 22 transports bytes + files.
        stream.close()
    }
}

// MARK: - MCNearbyServiceBrowserDelegate

extension PeerServicePlugin: MCNearbyServiceBrowserDelegate {
    public func browser(_ browser: MCNearbyServiceBrowser, foundPeer peerID: MCPeerID, withDiscoveryInfo info: [String: String]?) {
        let id = peerKey(for: peerID)
        peersById[id] = peerID
        if stateById[id] == nil { stateById[id] = "found" }
        emitPeers()
    }

    public func browser(_ browser: MCNearbyServiceBrowser, lostPeer peerID: MCPeerID) {
        let id = peerKey(for: peerID)
        peersById.removeValue(forKey: id)
        stateById.removeValue(forKey: id)
        emitPeers()
    }
}

// MARK: - MCNearbyServiceAdvertiserDelegate

extension PeerServicePlugin: MCNearbyServiceAdvertiserDelegate {
    public func advertiser(
        _ advertiser: MCNearbyServiceAdvertiser,
        didReceiveInvitationFromPeer peerID: MCPeerID,
        withContext context: Data?,
        invitationHandler: @escaping (Bool, MCSession?) -> Void
    ) {
        let id = peerKey(for: peerID)
        peersById[id] = peerID
        let inviteId = UUID().uuidString
        pendingInvites[inviteId] = invitationHandler
        inviteToPeer[inviteId] = peerID
        invitesHandler.send([
            "id": inviteId,
            "peerId": id,
            "peerName": peerID.displayName,
        ])
    }
}

// MARK: - SinkHandler

final class SinkHandler: NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
    func send(_ value: Any) {
        DispatchQueue.main.async { [weak self] in self?.sink?(value) }
    }
}

File: android/app/src/main/kotlin/com/example/noti_notes_app/PeerServicePlugin.kt
package com.example.noti_notes_app

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.google.android.gms.nearby.Nearby
import com.google.android.gms.nearby.connection.AdvertisingOptions
import com.google.android.gms.nearby.connection.ConnectionInfo
import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback
import com.google.android.gms.nearby.connection.ConnectionResolution
import com.google.android.gms.nearby.connection.ConnectionsClient
import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo
import com.google.android.gms.nearby.connection.DiscoveryOptions
import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback
import com.google.android.gms.nearby.connection.Payload
import com.google.android.gms.nearby.connection.PayloadCallback
import com.google.android.gms.nearby.connection.PayloadTransferUpdate
import com.google.android.gms.nearby.connection.Strategy
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Bridges spec 22 PeerService to Google Nearby Connections.
 *
 * Channels mirror the iOS implementation in `ios/Runner/PeerServicePlugin.swift`
 * and the Dart [ChannelPeerService] consumer.
 */
class PeerServicePlugin(private val context: Context, messenger: BinaryMessenger) {

    private val main = Handler(Looper.getMainLooper())
    private val client: ConnectionsClient = Nearby.getConnectionsClient(context)

    private val control = MethodChannel(messenger, "noti.peer/control")
    private val peersSink = SinkHandler().also {
        EventChannel(messenger, "noti.peer/peers").setStreamHandler(it)
    }
    private val invitesSink = SinkHandler().also {
        EventChannel(messenger, "noti.peer/invites").setStreamHandler(it)
    }
    private val payloadsSink = SinkHandler().also {
        EventChannel(messenger, "noti.peer/payloads").setStreamHandler(it)
    }
    private val transfersSink = SinkHandler().also {
        EventChannel(messenger, "noti.peer/transfers").setStreamHandler(it)
    }

    private var serviceType: String = "noti-share"
    private var displayName: String = "noti"
    private var advertising = false
    private var discovering = false

    /** endpointId -> friendly display name reported by remote */
    private val peerNames = ConcurrentHashMap<String, String>()

    /** endpointId -> "found" | "inviting" | "accepting" | "connected" | "disconnected" */
    private val peerStates = ConcurrentHashMap<String, String>()

    /** inviteId -> endpointId */
    private val pendingInvites = ConcurrentHashMap<String, String>()

    /** payloadId (long, as String) -> transferId (uuid) */
    private val transferIdsByPayload = ConcurrentHashMap<Long, String>()
    private val totalsByPayload = ConcurrentHashMap<Long, Long>()
    private val payloadEndpoint = ConcurrentHashMap<Long, String>()
    private val payloadDirection = ConcurrentHashMap<Long, String>()

    init {
        control.setMethodCallHandler { call, result -> handle(call, result) }
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "start" -> start(call, result)
                "stop" -> stop(result)
                "invite" -> invite(call, result)
                "acceptInvite" -> acceptInvite(call, result)
                "rejectInvite" -> rejectInvite(call, result)
                "sendBytes" -> sendBytes(call, result)
                "sendFile" -> sendFile(call, result)
                "cancelTransfer" -> cancelTransfer(call, result)
                "disconnect" -> disconnect(call, result)
                else -> result.notImplemented()
            }
        } catch (t: Throwable) {
            result.error("native_error", t.message, null)
        }
    }

    private fun start(call: MethodCall, result: MethodChannel.Result) {
        val role = call.argument<String>("role") ?: "both"
        displayName = call.argument<String>("displayName") ?: "noti"
        serviceType = call.argument<String>("serviceType") ?: "noti-share"

        if (role == "advertise" || role == "both") {
            client.startAdvertising(
                displayName,
                serviceType,
                connectionLifecycle,
                AdvertisingOptions.Builder().setStrategy(Strategy.P2P_CLUSTER).build()
            )
            advertising = true
        }
        if (role == "discover" || role == "both") {
            client.startDiscovery(
                serviceType,
                endpointDiscovery,
                DiscoveryOptions.Builder().setStrategy(Strategy.P2P_CLUSTER).build()
            )
            discovering = true
        }
        emitPeers()
        result.success(null)
    }

    private fun stop(result: MethodChannel.Result) {
        if (advertising) client.stopAdvertising()
        if (discovering) client.stopDiscovery()
        client.stopAllEndpoints()
        advertising = false
        discovering = false
        peerNames.clear()
        peerStates.clear()
        pendingInvites.clear()
        transferIdsByPayload.clear()
        totalsByPayload.clear()
        payloadEndpoint.clear()
        payloadDirection.clear()
        peersSink.send(emptyList<Map<String, Any>>())
        result.success(null)
    }

    private fun invite(call: MethodCall, result: MethodChannel.Result) {
        val peerId = call.argument<String>("peerId")
            ?: return result.error("bad_args", "peerId missing", null)
        peerStates[peerId] = "inviting"
        emitPeers()
        client.requestConnection(displayName, peerId, connectionLifecycle)
            .addOnSuccessListener { result.success(null) }
            .addOnFailureListener { e -> result.error("invite_failed", e.message, null) }
    }

    private fun acceptInvite(call: MethodCall, result: MethodChannel.Result) {
        val inviteId = call.argument<String>("inviteId")
            ?: return result.error("bad_args", "inviteId missing", null)
        val endpointId = pendingInvites.remove(inviteId)
            ?: return result.error("no_invite", "Unknown invite id", null)
        peerStates[endpointId] = "accepting"
        emitPeers()
        client.acceptConnection(endpointId, payloadCallback)
            .addOnSuccessListener { result.success(null) }
            .addOnFailureListener { e -> result.error("accept_failed", e.message, null) }
    }

    private fun rejectInvite(call: MethodCall, result: MethodChannel.Result) {
        val inviteId = call.argument<String>("inviteId")
            ?: return result.error("bad_args", "inviteId missing", null)
        val endpointId = pendingInvites.remove(inviteId)
            ?: return result.error("no_invite", "Unknown invite id", null)
        client.rejectConnection(endpointId)
            .addOnSuccessListener { result.success(null) }
            .addOnFailureListener { e -> result.error("reject_failed", e.message, null) }
    }

    private fun sendBytes(call: MethodCall, result: MethodChannel.Result) {
        val peerId = call.argument<String>("peerId")
            ?: return result.error("bad_args", "peerId missing", null)
        val bytes = call.argument<ByteArray>("bytes")
            ?: return result.error("bad_args", "bytes missing", null)
        val transferId = UUID.randomUUID().toString()
        val payload = Payload.fromBytes(bytes)
        registerOutgoing(payload.id, transferId, peerId, bytes.size.toLong())
        client.sendPayload(peerId, payload)
            .addOnSuccessListener { result.success(transferId) }
            .addOnFailureListener { e -> result.error("send_failed", e.message, null) }
    }

    private fun sendFile(call: MethodCall, result: MethodChannel.Result) {
        val peerId = call.argument<String>("peerId")
            ?: return result.error("bad_args", "peerId missing", null)
        val path = call.argument<String>("path")
            ?: return result.error("bad_args", "path missing", null)
        val maxBytes = (call.argument<Int>("maxBytes") ?: Int.MAX_VALUE).toLong()
        val file = File(path)
        if (!file.exists()) return result.error("no_file", "File not found", null)
        if (file.length() > maxBytes) return result.error("too_large", "File exceeds cap", null)
        val pfd = context.contentResolver.openFileDescriptor(android.net.Uri.fromFile(file), "r")
            ?: return result.error("no_file", "Cannot open file", null)
        val transferId = UUID.randomUUID().toString()
        val payload = Payload.fromFile(pfd)
        registerOutgoing(payload.id, transferId, peerId, file.length())
        client.sendPayload(peerId, payload)
            .addOnSuccessListener {
                // Nearby copies the FD before sendPayload returns; close ours
                // to avoid leaking a descriptor per file send.
                try { pfd.close() } catch (_: Throwable) {}
                result.success(transferId)
            }
            .addOnFailureListener { e ->
                try { pfd.close() } catch (_: Throwable) {}
                result.error("send_failed", e.message, null)
            }
    }

    private fun cancelTransfer(call: MethodCall, result: MethodChannel.Result) {
        val transferId = call.argument<String>("transferId")
            ?: return result.error("bad_args", "transferId missing", null)
        val payloadId = transferIdsByPayload.entries.firstOrNull { it.value == transferId }?.key
            ?: return result.success(null)
        client.cancelPayload(payloadId)
        result.success(null)
    }

    private fun disconnect(call: MethodCall, result: MethodChannel.Result) {
        val peerId = call.argument<String>("peerId")
            ?: return result.error("bad_args", "peerId missing", null)
        client.disconnectFromEndpoint(peerId)
        peerStates[peerId] = "disconnected"
        emitPeers()
        result.success(null)
    }

    private fun registerOutgoing(payloadId: Long, transferId: String, peerId: String, total: Long) {
        transferIdsByPayload[payloadId] = transferId
        totalsByPayload[payloadId] = total
        payloadEndpoint[payloadId] = peerId
        payloadDirection[payloadId] = "send"
        emitTransfer(transferId, peerId, "send", "queued", 0, total)
    }

    private fun emitPeers() {
        val list = peerNames.map { (id, name) ->
            mapOf(
                "id" to id,
                "displayName" to name,
                "state" to (peerStates[id] ?: "found")
            )
        }
        peersSink.send(list)
    }

    private fun emitTransfer(
        transferId: String,
        peerId: String,
        direction: String,
        phase: String,
        bytes: Long,
        total: Long,
        error: String? = null,
    ) {
        val ev = mutableMapOf<String, Any>(
            "transferId" to transferId,
            "peerId" to peerId,
            "direction" to direction,
            "phase" to phase,
            "bytes" to bytes.toInt(),
            "total" to total.toInt(),
        )
        if (error != null) ev["error"] = error
        transfersSink.send(ev)
    }

    private val endpointDiscovery = object : EndpointDiscoveryCallback() {
        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
            peerNames[endpointId] = info.endpointName
            peerStates[endpointId] = "found"
            emitPeers()
        }

        override fun onEndpointLost(endpointId: String) {
            peerNames.remove(endpointId)
            peerStates.remove(endpointId)
            emitPeers()
        }
    }

    private val connectionLifecycle = object : ConnectionLifecycleCallback() {
        override fun onConnectionInitiated(endpointId: String, info: ConnectionInfo) {
            peerNames[endpointId] = info.endpointName
            // If we initiated, we already accept implicitly; otherwise emit invite.
            if (peerStates[endpointId] == "inviting") {
                client.acceptConnection(endpointId, payloadCallback)
            } else {
                val inviteId = UUID.randomUUID().toString()
                pendingInvites[inviteId] = endpointId
                invitesSink.send(
                    mapOf(
                        "id" to inviteId,
                        "peerId" to endpointId,
                        "peerName" to info.endpointName,
                    )
                )
            }
        }

        override fun onConnectionResult(endpointId: String, result: ConnectionResolution) {
            peerStates[endpointId] = if (result.status.isSuccess) "connected" else "disconnected"
            emitPeers()
        }

        override fun onDisconnected(endpointId: String) {
            peerStates[endpointId] = "disconnected"
            emitPeers()
        }
    }

    private val payloadCallback = object : PayloadCallback() {
        override fun onPayloadReceived(endpointId: String, payload: Payload) {
            val transferId = transferIdsByPayload.getOrPut(payload.id) { UUID.randomUUID().toString() }
            payloadEndpoint[payload.id] = endpointId
            payloadDirection[payload.id] = "receive"
            when (payload.type) {
                Payload.Type.BYTES -> {
                    val bytes = payload.asBytes() ?: ByteArray(0)
                    payloadsSink.send(
                        mapOf(
                            "peerId" to endpointId,
                            "bytes" to bytes,
                        )
                    )
                    emitTransfer(transferId, endpointId, "receive", "completed", bytes.size.toLong(), bytes.size.toLong())
                }
                Payload.Type.FILE -> {
                    val uri = payload.asFile()?.asUri()
                    val path = uri?.path
                    if (path != null) {
                        payloadsSink.send(
                            mapOf(
                                "peerId" to endpointId,
                                "bytes" to ByteArray(0),
                                "filePath" to path,
                            )
                        )
                    }
                }
                else -> Unit
            }
        }

        override fun onPayloadTransferUpdate(endpointId: String, update: PayloadTransferUpdate) {
            val transferId = transferIdsByPayload[update.payloadId] ?: return
            val total = totalsByPayload[update.payloadId] ?: update.totalBytes
            val direction = payloadDirection[update.payloadId] ?: "receive"
            val phase = when (update.status) {
                PayloadTransferUpdate.Status.IN_PROGRESS -> "inProgress"
                PayloadTransferUpdate.Status.SUCCESS -> "completed"
                PayloadTransferUpdate.Status.CANCELED -> "cancelled"
                PayloadTransferUpdate.Status.FAILURE -> "failed"
                else -> "inProgress"
            }
            emitTransfer(transferId, endpointId, direction, phase, update.bytesTransferred, total)
            if (phase != "inProgress") {
                transferIdsByPayload.remove(update.payloadId)
                totalsByPayload.remove(update.payloadId)
                payloadEndpoint.remove(update.payloadId)
                payloadDirection.remove(update.payloadId)
            }
        }
    }

    inner class SinkHandler : EventChannel.StreamHandler {
        private var sink: EventChannel.EventSink? = null
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            sink = events
        }
        override fun onCancel(arguments: Any?) {
            sink = null
        }
        fun send(value: Any) {
            main.post { sink?.success(value) }
        }
    }
}
